tests/test_koji_util.py

```diff
@@ -96,21 +100,64 @@ def test_koji_login_ssl(self, tmpdir, proxyuser, serverca):
 
 class TestCreateKojiSession(object):
     def test_create_simple_session(self):
-        url = 'https://koji-hub-url.com'
+        url = 'https://example.com'
         session = flexmock()
 
         (flexmock(koji_util.koji).should_receive('ClientSession').with_args(
             url, opts={'krb_rdns': False}).and_return(session))
-        assert create_koji_session(url) == session
+        assert create_koji_session(url)._wrapped_session == session
+
+    @pytest.mark.parametrize(('ssl_session'), [
+        (True, False),
+    ])
+    def test_create_authenticated_session(self, tmpdir, ssl_session):
+        url = 'https://example.com'
+        args = {}
 
-    def test_create_authenticated_session(self):
-        url = 'https://koji-hub-url.com'
         session = flexmock()
-        session.should_receive('krb_login').once().and_return(True)
+        if ssl_session:
+            args['ssl_certs_dir'] = str(tmpdir)
+            session.should_receive('ssl_login').and_return(True)
+        else:
+            session.should_receive('krb_login').once().and_return(True)
 
         (flexmock(koji_util.koji).should_receive('ClientSession').with_args(
             url, opts={'krb_rdns': False}).and_return(session))
-        assert create_koji_session(url, {}) == session
+        assert create_koji_session(url, args)._wrapped_session == session
+
+    @pytest.mark.parametrize(('auth_type', 'auth_args'), [
```

Comment:
Covers the same branch in auth logic as test case above; I'd rather this just focused on the one behavior of new interest, the exception retry.

Reply:
The session wrapper code needs to handle 3 cases: genuine exception such as a KeyError, ConnectionError that times out, and momentary ConnectionError that clears on a retry. And it needs to do it for ssl and krb logins.

The clears on a retry eventually ends up in the same place as a successful session, but I think there's merit in testing both. Both to get 100% coverage and to show how the code is supposed to work.